miniscript: convert non-critical asserts to CHECK_NONFATAL

darosior opened the pull request:
The Miniscript code contains assertions to prevent ending up in an insane state or prevent UB, but also to enforce logical invariants. For the latter it is not necessary to crash the program if they are broken. Raising an exception suffices, especially as this code is often called through the RPC interface which can in turn handle the exception and the user can report it to developers.

This revives #28678 from Pieter Wuille.

darosior committed:
miniscript: convert non-critical asserts to CHECK_NONFATAL

The Miniscript code contains assertions to prevent ending up in an insane state or prevent UB, but
also to enforce logical invariants. For the latter it is not necessary to crash the program if they
are broken. Raising an exception suffices, especially as this code is often called through the RPC
interface which can in turn handle the exception and the user can report it to developers.

This is based on previous work from Pieter Wuille.

brunoerg commented:
Concept ACK

pythcoiner commented on src/script/miniscript.h:
```diff
@@ -659,7 +659,8 @@ struct Node {
         }
         // The final remaining results element is the root result, return it.
-        assert(results.size() == 1);
+        assert(results.size() >= 1);
```
why `>=` if then you check against `==` ?

darosior replied:
The `assert` is a looser check. If this condition doesn't hold it will trigger UB on the line right after so we'd rather crash. The `CHECK_NONFATAL` check is stricter and as long as the assert holds only applies to the logic, so it's not necessary to crash the program. Raising an exception to signal the bug to the user is sufficient.

hodlinator reviewed (commented):
Concept ACK ff0194a7ce9dabf1b31b64ca584e45840dce8141

Why aren't these converted to use `CHECK_NONFATAL`?
https://github.com/bitcoin/bitcoin/blob/ff0194a7ce9dabf1b31b64ca584e45840dce8141/src/script/miniscript.h#L1011-L1013

https://github.com/bitcoin/bitcoin/blob/ff0194a7ce9dabf1b31b64ca584e45840dce8141/src/script/miniscript.h#L1179-L1181

darosior commented:
> Why aren't these converted to use `CHECK_NONFATAL`?

Same reason as [here](https://github.com/bitcoin/bitcoin/pull/31727#discussion_r1929121645). I kept asserts where breaking the invariant would lead to UB. In the examples you point out it would read out of the `sats` vector bounds.

hodlinator commented:
> Same reason as [here](https://github.com/bitcoin/bitcoin/pull/31727#discussion_r1929121645). I kept asserts where breaking the invariant would lead to UB. In the examples you point out it would read out of the `sats` vector bounds.

Okay, and there is no way *user input* from RPCs etc would currently be able to trigger the asserts?

> Actually this should be a strict check, not an inferior or equal to!

Ah, didn't realize, good catch indeed, `=` should be dropped.

kevkevinpal commented:
Concept ACK

lightly reviewed the code and looks good to me

sedited reviewed (approved):
ACK ff0194a7ce9dabf1b31b64ca584e45840dce8141

brunoerg reviewed (approved):
code review ACK ff0194a7ce9dabf1b31b64ca584e45840dce8141

hodlinator reviewed (approved):
ACK ff0194a7ce9dabf1b31b64ca584e45840dce8141

Makes sense to not crash when encountering invalid miniscript.

#### Testing

Inserted `CHECK_NONFATAL(false)` in two separate touched functions to see how RPCs behave:
```
₿ build/bin/bitcoin-cli -regtest deriveaddresses "wsh(1)#mrg7xj7p"
error code: -1
error message:
Internal bug detected: false
../src/script/miniscript.cpp:20 (SanitizeType)
Bitcoin Core v28.99.0-ff0194a7ce9d-dirty
Please report this issue here: https://github.com/bitcoin/bitcoin/issues
```
Similar for:
```
../src/script/miniscript.cpp:41 (ComputeType)
```

<details><summary>

#### Following up on my [unanswered message](https://github.com/bitcoin/bitcoin/pull/31727#issuecomment-2621725687):

</summary>

> > Same reason as [here](https://github.com/bitcoin/bitcoin/pull/31727#discussion_r1929121645). I kept asserts where breaking the invariant would lead to UB. In the examples you point out it would read out of the `sats` vector bounds.
> 
> Okay, and there is no way _user input_ from RPCs etc would currently be able to trigger the asserts?

~~(I acknowledge you drew the line at UB, but out of bounds access fits that definition AFAIK).~~

Seems like user input like excessive thresh `k`-value is guarded against at least to some extent here while string parsing:

https://github.com/bitcoin/bitcoin/blob/ff0194a7ce9dabf1b31b64ca584e45840dce8141/src/script/miniscript.h#L2125-L2133

So the `assert`s referenced in https://github.com/bitcoin/bitcoin/pull/31727#pullrequestreview-2577746602 will typically run after that guard. Meaning there's little point in converting those checks to `CHECK_NONFATAL`.

Tested:
```
₿ build/bin/bitcoin-cli -regtest deriveaddresses "wsh(thresh(1,pk(tpubDFkN51vYF36W4Yfn3wGv5fpmRo3ok7vZZjc1gmRJjumq33L776e6GkP4HGdCVjDqYiBahXCrXQKja8aUZ2xovQNS8WkF46MdY7TLHJLYD7H/0/0)))#qzdpxsqa"
[
  "bcrt1q2kwrzducejduqjf36zmx6qzq29l3gc5sca7z2fqcyshm5jr0v4vq7erjda"
]

₿ build/bin/bitcoin-cli -regtest deriveaddresses "wsh(thresh(2,pk(tpubDFkN51vYF36W4Yfn3wGv5fpmRo3ok7vZZjc1gmRJjumq33L776e6GkP4HGdCVjDqYiBahXCrXQKja8aUZ2xovQNS8WkF46MdY7TLHJLYD7H/0/0)))#zx6ca7mc"
error code: -5
error message:
A function is needed within P2WSH
```
Not the clearest error message, but at least an error.

> > Actually this should be a strict check, not an inferior or equal to!
> 
> Ah, didn't realize, good catch indeed, `=` should be dropped.

Would still be nice if this were addressed, but bug wasn't caused by this PR, so could be a follow-up.

Edit: Stroke out inverted logic.

</details>

hodlinator commented:
@darosior discovered (https://github.com/bitcoin/bitcoin/pull/31727#issuecomment-2619342125) in response to https://github.com/bitcoin/bitcoin/pull/31727#pullrequestreview-2577746602:
> Ah, didn't realize, good catch indeed, `=` should be dropped.

This is now fixed together with some more cases in follow-up #32255.

darosior commented:
Thanks for picking it up @hodlinator.

-------- Original Message --------
On 4/12/25 4:06 AM, hodlinator  wrote:

> ***@***.***(https://github.com/darosior) discovered ([#31727 (comment)](https://github.com/bitcoin/bitcoin/pull/31727#issuecomment-2619342125)) in response to [#31727 (review)](https://github.com/bitcoin/bitcoin/pull/31727#pullrequestreview-2577746602):
>
>> Ah, didn't realize, good catch indeed, = should be dropped.
>
> This is now fixed together with some more cases in follow-up [#32255](https://github.com/bitcoin/bitcoin/pull/32255).
>
> —
> Reply to this email directly, [view it on GitHub](https://github.com/bitcoin/bitcoin/pull/31727#issuecomment-2798727539), or [unsubscribe](https://github.com/notifications/unsubscribe-auth/AFLK3F7LAMHJFREVZJC7WRD2ZDCORAVCNFSM6AAAAABVXXMUSOVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDOOJYG4ZDONJTHE).
> You are receiving this because you were mentioned.Message ID: ***@***.***>
>
> hodlinator left a comment [(bitcoin/bitcoin#31727)](https://github.com/bitcoin/bitcoin/pull/31727#issuecomment-2798727539)
>
> ***@***.***(https://github.com/darosior) discovered ([#31727 (comment)](https://github.com/bitcoin/bitcoin/pull/31727#issuecomment-2619342125)) in response to [#31727 (review)](https://github.com/bitcoin/bitcoin/pull/31727#pullrequestreview-2577746602):
>
>> Ah, didn't realize, good catch indeed, = should be dropped.
>
> This is now fixed together with some more cases in follow-up [#32255](https://github.com/bitcoin/bitcoin/pull/32255).
>
> —
> Reply to this email directly, [view it on GitHub](https://github.com/bitcoin/bitcoin/pull/31727#issuecomment-2798727539), or [unsubscribe](https://github.com/notifications/unsubscribe-auth/AFLK3F7LAMHJFREVZJC7WRD2ZDCORAVCNFSM6AAAAABVXXMUSOVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDOOJYG4ZDONJTHE).
> You are receiving this because you were mentioned.Message ID: ***@***.***>
